score.go

```diff
@@ -0,0 +1,952 @@
+	}
+
+	// check P1
+	if p.TimeInMeshWeight < 0 {
+		return fmt.Errorf("invalid TimeInMeshWeight; must be positive (or 0 to disable)")
+	}
+	if p.TimeInMeshWeight != 0 && p.TimeInMeshQuantum <= 0 {
+		return fmt.Errorf("invalid TimeInMeshQuantum; must be positive")
+	}
+	if p.TimeInMeshWeight != 0 && p.TimeInMeshCap <= 0 {
+		return fmt.Errorf("invalid TimeInMeshCap; must be positive")
+	}
+
+	// check P2
+	if p.FirstMessageDeliveriesWeight < 0 {
+		return fmt.Errorf("invallid FirstMessageDeliveriesWeight; must be positive (or 0 to disable)")
+	}
+	if p.FirstMessageDeliveriesWeight != 0 && (p.FirstMessageDeliveriesDecay <= 0 || p.FirstMessageDeliveriesDecay >= 1) {
+		return fmt.Errorf("invalid FirstMessageDeliveriesDecay; must be between 0 and 1")
+	}
+	if p.FirstMessageDeliveriesWeight != 0 && p.FirstMessageDeliveriesCap <= 0 {
+		return fmt.Errorf("invalid FirstMessageDeliveriesCap; must be positive")
+	}
+
+	// check P3
+	if p.MeshMessageDeliveriesWeight > 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesWeight; must be negative (or 0 to disable)")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && (p.MeshMessageDeliveriesDecay <= 0 || p.MeshMessageDeliveriesDecay >= 1) {
+		return fmt.Errorf("invalid MeshMessageDeliveriesDecay; must be between 0 and 1")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesCap <= 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesCap; must be positive")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesThreshold <= 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesThreshold; must be positive")
+	}
+	if p.MeshMessageDeliveriesWindow < 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesWindow; must be non-negative")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesActivation < time.Second {
+		return fmt.Errorf("invalid MeshMessageDeliveriesActivation; must be at least 1s")
+	}
+
+	// check P3b
+	if p.MeshFailurePenaltyWeight > 0 {
+		return fmt.Errorf("invalid MeshFailurePenaltyWeight; must be negative (or 0 to disable)")
+	}
+	if p.MeshFailurePenaltyWeight != 0 && (p.MeshFailurePenaltyDecay <= 0 || p.MeshFailurePenaltyDecay >= 1) {
+		return fmt.Errorf("invalid MeshFailurePenaltyDecay; must be between 0 and 1")
+	}
+
+	// check P4
+	if p.InvalidMessageDeliveriesWeight >= 0 {
+		return fmt.Errorf("invalid InvalidMessageDeliveriesWeight; must be negative")
+	}
+	if p.InvalidMessageDeliveriesDecay <= 0 || p.InvalidMessageDeliveriesDecay >= 1 {
+		return fmt.Errorf("invalid InvalidMessageDeliveriesDecay; must be between 0 and 1")
+	}
+
+	return nil
+}
+
+// router interface
+func (ps *peerScore) Start(gs *GossipSubRouter) {
+	if ps == nil {
+		return
+	}
+
+	ps.msgID = gs.p.msgID
+	ps.host = gs.p.host
+	go ps.background(gs.p.ctx)
+}
+
+func (ps *peerScore) Score(p peer.ID) float64 {
+	if ps == nil {
+		return 0
+	}
+
+	ps.Lock()
+	defer ps.Unlock()
+
+	return ps.score(p)
+}
+
+func (ps *peerScore) score(p peer.ID) float64 {
```

Comment:
Eventually, we might find that we'd benefit from caching the score, and recalculating it in batch every N ticks (resolution), under a global lock. Recalculating the score on access may be heavy and may introduce minor skews that may create subtle unfairness across peers based on access time.

Reply:
I thought about it, but I think it's premature to do it now.